src/shogun/machine/GLM.h

```diff
@@ -0,0 +1,293 @@
+        //TODO GAMMA,
+        //TODO SOFTPLUS,
+        //TODO PROBIT,
+        POISSON
+	};
+
+class DotFeatures;
+class Features;
+class RegressionLabels;
+class GLMCostFunction;
+
+/** @brief Class for estimating regularized generalized linear models (GLM).
+ *   The regularized GLM minimizes the penalized negative log likelihood:
+ *
+ *  This uses Elastic-net penalty which defaults to the ridge penalty when
+ *  alpha = 0 and defaults to the lasso penalty when alpha = 1.
+ *
+ * */
+class GLM : public RandomMixin<IterativeMachine<LinearMachine>>
+{
+ 	public:
+		friend class GLMCostFunction;
+		
+		/** default constructor */
+		GLM();
+
+ 		GLM(GLM_DISTRIBUTION distribution, float64_t alpha, float64_t lambda, float64_t learning_rate, int32_t max_iterations, float64_t tolerance, float64_t eta);
+
+ 		/** destructor */
+ 		~GLM() override {}
+
+		/** apply linear machine to data
+		 * for regression problem
+		 *
+		 * @param data (test)data to be classified
+		 * @return classified labels
+		 */
+		std::shared_ptr<RegressionLabels> apply_regression(std::shared_ptr<Features> data=NULL) override;
+
+ 		/** Returns the name of the SGSerializable instance.  It MUST BE
+ 		 *  the CLASS NAME without the prefixed `C'.
+ 		 *
+ 		 * @return name of the SGSerializable
+ 		 */
+ 		const char* get_name() const override { return "GLM"; }
+
+		void set_tau(SGMatrix<float64_t> tau);
+
+		SGMatrix<float64_t> get_tau();
+
+	protected:
+
+		void init_model(const std::shared_ptr<Features>& data);
+		
+		void iteration() override;
+
+	protected:
+
+		/** Distribution type */
+		GLM_DISTRIBUTION distribution = POISSON;
+
+		/** a threshold parameter that linearizes the exp() function above eta. */
+		float64_t m_eta = 2.0;
+
+		/** regularization parameter :math:`\\lambda` of penalty term. */
+		float64_t m_lambda = 0.1;
+
+		/** the weighting between L1 penalty and L2 penalty term of the loss function. */
+		float64_t m_alpha = 0.5;
+
+		/** convergence threshold or stopping criteria. Optimization loop will stop when relative change in parameter norm is below the threshold. */
+		float64_t m_tolerance = 1e-6;
+
+		bool m_compute_bias = true;
+
+		std::shared_ptr<GradientDescendUpdater> m_gradient_updater;
+
+		std::shared_ptr<ConstLearningRate> m_learning_rate;
+
+		std::shared_ptr<ElasticNetPenalty> m_penalty;
+
+		std::shared_ptr<GLMCostFunction> m_cost_function;
+};
+
+class GLMCostFunction: public FirstOrderCostFunction
+{
+public:
+
+	friend class GLM;
+
+	GLMCostFunction()
+	{
+
+	}
+
+	GLMCostFunction(const std::shared_ptr<GLM>&obj)
+	{
+		if(m_obj != obj)
+			m_obj=obj;
+	}
+
+	void set_target(const std::shared_ptr<GLM>&obj)
+	{
+		if(m_obj != obj)
+			m_obj=obj;
+	}
+
+	void unset_target()
+	{
+		m_obj=NULL;
+	}
+
+	virtual float64_t get_cost()
+	{
+		//TODO
+		return 0.0;
+	}
+
+	virtual SGVector<float64_t> obtain_variable_reference()
+	{
+		require(m_obj,"Object not set");
+		return m_obj->m_w;
+	}
+
+	virtual SGVector<float64_t> get_gradient()
```

Comment:
I just realized this. Does this class really need to keep a pointer to a `GLM` object? Wouldn't be possible to change its design such to have something like
```c++
virtual SGVector<float64_t> get_gradient(const Features &X, const Labels &y, const SGVector& weights, const float64_t bias)
```
Then, inside the GLM object, you will just call
```c++
get_gradient(m_features->get_computed_dot_feature_matrix(), labels, m_weigths, m_bias)
```
We will be able to decouple the `GLMCostFunction` from the GLM object and, as a bonus, you will be able to write a unit test for the gradients more easily (since `get_gradient()` is public here).

Reply:
Oh okay, so I should pass all the variables as parameters. Should I make the functions static?

Reply:
I would say no. I would like to have in the future multiple classes which stem from `GLMCostFunction` which represents single distributions (e.g., `GLMPoisson`, `GLMBinomial`, etc.) and each of them will have to override `get_gradient()` to implement its own gradient.

Reply:
Hmm right, this makes sense, So I should implement get_gradient the way I have implemented `compute_z(const SGMatrix<float64_t> X, const SGVector<float64_t> w, const float64_t bias)` or `non_linearity(const SGVector<float64_t> z)`.

And I should make this get_gradient() public as well so that we can test it directly using a unit test.
I'm on it.

Reply:
Yes! Focus just on changing `GLMCostFunction`'s methods for now. We will refactor into several different classes later.

Reply:
Should I pass features and lables as parameters? Or SGMatrix and SGVector?

Reply:
Your choice. Try to keep it consistent with the other methods though.

Reply:
Okay so I figured out why i hadn't done this in the first place.
GLMCostFunction is derived from FirstOrderCostFunction which has the pure virtual function `virtual SGVector<float64_t> get_gradient()=0;`
So, I can't add parameters.
but there is an option to skip the inheritance all together and just write GLMCostFunction as an independent class. Is that okay?

Reply:
Yes, do it.

Reply:
You would need it to be derived from `FirstOrderCostFunction` so that it can be exposed to the interfaces (not sure that is necessary)?

Reply:
But, then how do we separately add a unit test for the get_gradient() function if it is supposed to be protected in the `FirstOrderCostFunction`

Reply:
I think `FirstOrderCostFunction` was written for a different purpose. I would keep `GLMCostFunction` as a separate class. It is not needed to expose it through the interfaces.